File: testers/avro/java-spring/generated-and-checked-in/com/example/events/InvoicePublisher.java
package com.example.events;

import java.util.concurrent.CompletableFuture;
import org.springframework.kafka.core.KafkaTemplate;
import org.springframework.kafka.support.SendResult;
import org.springframework.stereotype.Service;

/** Type-safe event publisher for invoice topic */
@Service
public record InvoicePublisher(KafkaTemplate<String, Invoice> kafkaTemplate, String topic) {
  public InvoicePublisher(KafkaTemplate<String, Invoice> kafkaTemplate) {
    this(kafkaTemplate, "invoice");
  }

  public InvoicePublisher withKafkaTemplate(KafkaTemplate<String, Invoice> kafkaTemplate) {
    return new InvoicePublisher(kafkaTemplate, topic);
  }

  public InvoicePublisher withTopic(String topic) {
    return new InvoicePublisher(kafkaTemplate, topic);
  }

  /** Publish a Invoice event */
  public CompletableFuture<SendResult<String, Invoice>> publish(String key, Invoice event) {
    return kafkaTemplate.send(topic, key, event);
  }
}

File: testers/avro/scala-cats/generated-and-checked-in/com/example/events/InvoicePublisher.scala
package com.example.events

import cats.effect.IO
import fs2.kafka.KafkaProducer
import fs2.kafka.ProducerRecord
import fs2.kafka.ProducerRecords
import fs2.kafka.ProducerResult
import scala.annotation.unused

@unused
/** Type-safe event publisher for invoice topic */
case class InvoicePublisher(
  kafkaTemplate: KafkaProducer[IO, String, Invoice],
  topic: String = "invoice"
) {
  /** Publish a Invoice event */
  def publish(
    key: String,
    event: Invoice
  ): IO[ProducerResult[String, Invoice]] = {
    return kafkaTemplate.produce(ProducerRecords.one(ProducerRecord(topic, key, event))).flatten
  }
}